### PR TITLE
Optimize syncing logic to handle large outpoint sets

### PR DESCRIPTION
**Pull request by rileystephens28**
- Use map for Qi wallet available outpoints
- Improve efficiency of syncing request batching 
- Minor bug fixes

### CODE REVIEW

**Review by alejoacosta74 — CHANGES_REQUESTED** [review]
Qi address derivation tests are failing. Mainly due to :
- `"lastSyncBlock"` property not found in `QiAddressInfo` (this is expected, unit test data needs to be updated)
- Address status expected is `"UNUSED"` but got `"UNKNOWN"`. Is this expected?

```
 [ fail(1): derives external addresses correctly - External address mismatch for zone 0x00, expected: {"pubKey":"0x02be08725a5cef5a8eb4bf27b3f0e4d256663e44c25db8e9705aff27ee322f3384","address":"0x009f1545923a5A1052Aa162F858e2b925863Cd3D","index":385,"change":false,"zone":"0x00","status":"UNUSED","derivationPath":"BIP44:external","account":0}, derived: {"address":"0x009f1545923a5A1052Aa162F858e2b925863Cd3D","pubKey":"0x02be08725a5cef5a8eb4bf27b3f0e4d256663e44c25db8e9705aff27ee322f3384","account":0,"index":385,"change":false,"zone":"0x00","status":"UNKNOWN","derivationPath":"BIP44:external","lastSyncedBlock":null} ]
 ```

File: src/wallet/qi-hdwallet.ts [hunk 1]
```diff
@@ -252,6 +254,21 @@ export class QiHDWallet extends AbstractHDWallet<QiAddressInfo> {
...
+        //         account,
+        //         index: 0,
+        //         change: isChange,
+        //         zone,
+        //         status: AddressStatus.UNKNOWN,
+        //         derivationPath: 'BIP44:external',
+        //         lastSyncedBlock: null,
+        //     };
+        // }
+
```
[the remark this answers is not in the document]
Comment (rileystephens28) [reply src/wallet/qi-hdwallet.ts:271]: done

File: src/wallet/qi-hdwallet.ts [hunk 2]
```diff
@@ -252,6 +254,21 @@ export class QiHDWallet extends AbstractHDWallet<QiAddressInfo> {
...
         const lastIndex = this._findLastUsedIndex(addresses, account, zone);
         const addressNode = this.deriveNextAddressNode(account, lastIndex + 1, zone, isChange);
         const newAddressInfo = this._createAndStoreQiAddressInfo(addressNode, account, zone, isChange);
```
[the remark this answers is not in the document]
Comment (rileystephens28) [reply src/wallet/qi-hdwallet.ts:274]: implemented

**Review by rileystephens28 — COMMENTED** [review]
> Qi address derivation tests are failing. Mainly due to :
> - `"lastSyncBlock"` property not found in `QiAddressInfo` (this is expected, unit test data needs to be updated)
> - Address status expected is `"UNUSED"` but got `"UNKNOWN"`. Is this expected?
> 
> ```
>  [ fail(1): derives external addresses correctly - External address mismatch for zone 0x00, expected: {"pubKey":"0x02be08725a5cef5a8eb4bf27b3f0e4d256663e44c25db8e9705aff27ee322f3384","address":"0x009f1545923a5A1052Aa162F858e2b925863Cd3D","index":385,"change":false,"zone":"0x00","status":"UNUSED","derivationPath":"BIP44:external","account":0}, derived: {"address":"0x009f1545923a5A1052Aa162F858e2b925863Cd3D","pubKey":"0x02be08725a5cef5a8eb4bf27b3f0e4d256663e44c25db8e9705aff27ee322f3384","account":0,"index":385,"change":false,"zone":"0x00","status":"UNKNOWN","derivationPath":"BIP44:external","lastSyncedBlock":null} ]
>  ``` 

Yes, this is intended. I just address status from "UNUSED" to "UNKNOWN" immediately after derivation as this better represents the status.

**Review by alejoacosta74 — APPROVED** [review]
looks good

**Pull request merged**